app/models/shipit/deploy_spec/npm_discovery.rb

```diff
@@ -79,10 +85,84 @@ def package_version
         JSON.parse(package_json.read)['version']
       end
 
+      def publish_config
+        JSON.parse(package_json.read)['publishConfig']
```

Comment:
This could fail in many ways (and I've seen it fail in prod):

  - File could contain invalid JSON
  - The JSON document could be missing that key

Also you are re-reading and re-parsing the file in many places, it's quite wastful. Please factor this to read and parse it once and memoize it.

Reply:
> File could contain invalid JSON

If a npm library contains invalid json from a `package.json` it would have been flagged well before this line. (ie. CI from a failed `yarn install`) .

> The JSON document could be missing that key

All references of `publish_config` fail appropriately if `publish_config.blank?`. This is covered in tests too. 

I've memoized all the file reads introduced from this PR in https://github.com/Shopify/shipit-engine/pull/767/commits/94b1b9057c6ceb01696b758a458fddf7891e7cd7.